### ISSUE TITLE
Using Different Motifs

### ISSUE THREAD

**Issue by KyleFerchen** [morris-lab/CellOracle#9]
Is it possible to use custom motifs for the scan?

It seems the gimmemotifs package links to an older version of the CisBP, and I was wondering if I could use version 2 of CisBP.

**Comment by KenjiKamimoto-wustl122** [comment]
Thank you for using celloracle!
Current version of CellOracle (0.3.2) does not support custom motifs.
But we are planning to update celloracle to be able to use custom motifs.
We will also update out tutorial to explain how to do that.
The next update is coming soon. Please wait for a while.

**Comment by KyleFerchen** [comment]
I worked this in by modifying the scan() method for the TFinfo class in the "celloracle/motif_analysis/tfinfo_core.py" file to accept a custom list of motifs read in through the gimme motifs package. Here is the code for the alternative "scan" class method definition I used, which includes a new parameter for a list of motifs. If none is provided, it defaults back to the original motifs that the gimmemotifs package loads:

    def scan(self, background_length=200, fpr=0.02, n_cpus=-1, verbose=True, motifs=None):
        """
        Scan DNA sequences searching for TF binding motifs.

        Args:
           background_length (int): background length. This is used for the calculation of the binding score.

           fpr (float): False positive rate for motif identification.

           n_cpus (int): number of CPUs for parallel calculation.

           verbose (bool): Whether to show a progress bar.

           motifs (list): a list of gimmemotifs motifs, will revert to default_motifs() if None

        """


        self.fpr = fpr
        self.background_length = background_length
        print("initiating scanner ...")
        ## 1. initialilze scanner  ##
        # load motif
        if not motifs:
            motifs = default_motifs()

        # initialize scanner
        s = Scanner(ncpus=n_cpus)

        # set parameters
        s.set_motifs(motifs)
        s.set_background(genome=self.ref_genome, length=background_length)
        #s.set_background(genome="mm9", length=400)
        s.set_threshold(fpr=fpr)

        ## 2. motif scan ##
        print("getting DNA sequences ...")
        target_sequences = peak2fasta(self.all_peaks, self.ref_genome)
        print("scanning motifs ...")
        self.scanned_df = scan_dna_for_motifs(s, motifs, target_sequences, verbose)

        self.__addLog("scanMotifs")


For generating the gimmemotifs list of motifs, the read_motifs() function (from gimmemotifs.motif import read_motifs). This function will read in the motifs from a .pfm file which structures the motifs like this:

\>M09623_2.00
0.0010	0.1660	0.3840	0.4480
0.0240	0.9380	0.0330	0.0050
0.9310	0.0310	0.0330	0.0050
0.1090	0.2560	0.4770	0.1570
0.1590	0.3450	0.3680	0.1280
0.2260	0.1780	0.5340	0.0620
0.6560	0.0240	0.1710	0.1490
0.0010	0.4080	0.5900	0.0010
0.0010	0.7550	0.0500	0.1940
0.1730	0.0100	0.8030	0.0140
0.0070	0.0140	0.0010	0.9780
0.0190	0.0010	0.9790	0.0010
0.9200	0.0010	0.0550	0.0240
0.0190	0.9710	0.0090	0.0010
\>M01730_2.00
0.3499	0.1468	0.2786	0.2248
0.3570	0.4174	0.1735	0.0521
0.1395	0.8279	0.0320	0.0006
0.0645	0.1262	0.2941	0.5153
0.5153	0.2941	0.1262	0.0645
0.0006	0.0320	0.8279	0.1395
0.0521	0.1735	0.4174	0.3570
0.2248	0.2786	0.1468	0.3499
\

>M04410_2.00
0.3038	0.2635	0.2975	0.1352
0.2033	0.6727	0.0507	0.0733
0.0000	0.8849	0.0000	0.1151
0.7225	0.2066	0.0000	0.0708
0.0168	0.9213	0.0000	0.0620
0.3735	0.0318	0.4735	0.1211
0.0064	0.9277	0.0292	0.0367
0.0420	0.9277	0.0000	0.0303
0.0469	0.7849	0.0158	0.1524
0.3136	0.3551	0.0566	0.2747
...


The read_motifs() function will filter the motifs using an annotation file in the same directory as the .pfm file. It has to have the same basename for the file, but with the extension ".motif2factors.txt". (Eg. "mouse_CIS-BP.motif2factors.txt" will filter "mouse_CIS-BP.pfm" when read in) I simply modeled the annotation file in the same way the gimmemotifs package structures their motif database files:
Motif	Factor	Evidence	Curated
M00111_2.00	Tfap2a	PBM	Y
M02768_2.00	Tfap2a	SELEX	Y
M02769_2.00	Tfap2a	SELEX	Y
M02770_2.00	Tfap2a	SELEX	Y
M08704_2.00	Tfap2a	Misc	Y
M00112_2.00	Tfap2b	PBM	Y
M08705_2.00	Tfap2b	Misc	Y
M00113_2.00	Tfap2c	PBM	Y
M08706_2.00	Tfap2c	Misc	Y
M00114_2.00	Tfap2e	PBM	Y
M00111_2.00	Tfap2d	PBM	N
M01659_2.00	Arid3b	PBM	Y
M00115_2.00	Arid3a	PBM	Y
M01660_2.00	Arid3a	PBM	Y
M02651_2.00	Arid3a	JASPAR	Y
M01661_2.00	Arid5b	PBM	Y
...

The important part is the Factor column, which will assign a direct link between the motif name and the transcription factor name. The motif name is used to link the motif to its pfm. The transcription factor name is needed for CellOracle to be able to connect the motif information to the gene expression data later.

For implementing it in the CellOracle workflow, I use read_motifs() function from gimmemotifs to first pull in the .pfm file that contains the motifs I want to use. There is an annotation file with the same basename and ".motif2factors.txt" that specifically has the motif names I want to use and their gene names. I filter the motif list to those that have a direct target. Then I use the modifed TFinfo.scan() method to use this motif list.


```
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pkg_resources
import gimmemotifs
import os, sys, shutil, importlib, glob
from tqdm.notebook import tqdm
from celloracle import motif_analysis as ma
from celloracle.utility import save_as_pickled_object
import celloracle as co

from gimmemotifs.motif import read_motifs

import genomepy

# Paths
output_path = "/media/kyle_storage/kyle_ferchen/grimes_lab_main/analysis/2020_06_02_modgmp_celloracle_cisbp2/output/03_tf_scan/"

gimme_motifs_pfm_file = "/home/kyle/anaconda3/envs/gimme_cisbp2/lib/python3.6/site-packages/data/motif_databases/mouse_CIS-BP.pfm"

peak_scan_space = "/media/kyle_storage/kyle_ferchen/grimes_lab_main/analysis/2020_06_02_modgmp_celloracle_cisbp2/output/02_peak_with_tss_anno/peak_file_50_coaccess.parquet"

peaks = pd.read_parquet(peak_scan_space)

# Filter out peaks that are too small
def getLength(x):
    a, b, c = x["peak_id"].split("_")
    return int(c) - int(b)

df = peaks.apply(lambda x: getLength(x), axis = 1)
peaks = peaks[df>=5]

ref_genome = "mm10"

# Instantiate TF info object
tfi = ma.TFinfo(peak_data_frame=peaks, ref_genome=ref_genome)

# Setup gimmemotifs motifs
mCisbp2 = read_motifs(gimme_motifs_pfm_file)
# Clean entries without direct connection
mClean = []
for motif in mCisbp2:
    if len(motif.factors['direct']) > 0:
        mClean.append(motif)

# Scan motifs
tfi.scan(fpr=0.05, verbose=True, n_cpus=8, motifs=mClean)

# Save tfinfo object
tfi.to_hdf5(file_path=output_path+"test.celloracle.tfinfo")

# Reset filtering
tfi.reset_filtering()

# tfi.filter_motifs_by_score(threshold=8, method='cumlative_score')
tfi.filter_motifs_by_score(threshold=4, method='indivisual_score')

```
That let me scan for motifs.

I also had to write a function to reinitialize the dic_motif2TFs value in the TFinfo object so that the tf gene names are linked to the motif ids:
```
def get_dic_motif2TFs(species, motifs):
    motif_names = [i.id for i in motifs]
    factors_direct = [i.factors["direct"] for i in motifs]
    factors_indirect = [i.factors["indirect"] for i in motifs]
    dic_motif2TFs = {}
    if species == "Mouse":
        for i in motifs:
            fcs = i.factors["direct"] + i.factors["indirect"]
            dic_motif2TFs[i.id] = [fa.capitalize() for fa in fcs]
    if species == "Human":
        for i in motifs:
            fcs = i.factors["direct"] + i.factors["indirect"]
            dic_motif2TFs[i.id] = [fa.upper() for fa in fcs]
    return dic_motif2TFs

tfi.dic_motif2TFs = get_dic_motif2TFs("Mouse", mClean)
```

After that, I could follow along with the rest of the CellOracle workflow, using the custom motifs.

**Comment by KenjiKamimoto-wustl122** [comment]
Hi Kyle, Thank you so much for sharing your codes!

It looks awesome! I will look into it and try to integrate it into celloracle!
I really appreciate your kind feedback! It helps us a lot.

**Comment by KenjiKamimoto-wustl122** [comment]
Hi Kyle, 

We made a major update in celloracle based on your suggestion. 
With the new version (celloracle>=0.5.0), we can select TF motif database easily. 

In the new version, we have mainly three options:
(1) We can import motif data from gimmemotifs dataset.
(2)We can load CisBP2 motifs that come with celloracle. I made motif data for several species using CisBP ver2 dataset. These CisBP2 data are built-in celloracle package.
(3) We can import custom motifs from XX.pfm file and XX.motifs2factors.txt file.

We also updated celloracle tutorial notebook.
https://morris-lab.github.io/CellOracle.documentation/notebooks/02_motif_scan/02_atac_peaks_to_TFinfo_with_celloracle_20200801.html

Thank you again for your great suggestion.
